File: src/codegen.rs
use std::io::{Result, Write};
use crate::{ast::{UnaryOp::*, BinaryOp::*}, ir::*};

pub fn write_asm(p: &IrProg, w: &mut impl Write) -> Result<()> {
  for g in &p.globs {
    writeln!(w, ".data")?;
    writeln!(w, "{}:", g.0)?;
    match g.1 {
      Ok(init) => writeln!(w, "  .word {}", init)?,
      Err(size) => writeln!(w, "  .zero {}", size * 4)?,
    }
    writeln!(w)?;
  }
  for f in &p.funcs {
    writeln!(w, ".global {}", f.name)?;
    writeln!(w, "{}:", f.name)?;
    // 函数的prologue部分，保存s0和ra寄存器的值，为局部变量开辟空间
    // 进入函数时，0(sp)的位置指向最后一个参数(如果存在的话)
    // 参数的id范围是0..param_cnt，局部变量的id范围是param_cnt..param_cnt + var_cnt，为了能让id简单地对应到偏移量，要让局部变量直接保存在参数后面这片区域
    writeln!(w, "  sw s0, -{}(sp)", (f.var_cnt + 1) * 4)?;
    writeln!(w, "  sw ra, -{}(sp)", (f.var_cnt + 2) * 4)?;
    writeln!(w, "  add s0, sp, {}", f.param_cnt * 4)?;
    writeln!(w, "  add sp, sp, -{}", (f.var_cnt + 2) * 4)?;
    // prologue结束后，栈的布局是这样的：
    //          <-- s0，访问参数0用-4(s0)
    // 参数0
    // 参数1
    // ..
    // 参数m
    // 局部变量0 <-- 访问它用-4 - 4 * m(s0)
    // 局部变量1
    // ...
    // 局部变量n
    // 保存的s0
    // 保存的ra <-- sp
    // <运算栈>
    //
    // 运算栈上可能会进行各种弹栈，压栈操作，但是栈指针始终不会越过prologue结束时的栈指针的位置，上面这些内容与运算栈互不干扰
    for s in &f.stmts {
      writeln!(w, "  # {:?}", s)?; // 输出一条注释，表示下面的汇编对应IR中的什么指令，方便调试
      match s {
        IrStmt::Const(x) => {
          writeln!(w, "  li t0, {}", x)?; // 将常数值保存到t0中
          writeln!(w, "  sw t0, -4(sp)")?; // 以下两句将t0的值压入栈中
          writeln!(w, "  add sp, sp, -4")?;
        }
        IrStmt::Unary(op) => {
          // 逻辑上这里应该分成三部：弹出运算数，运算，压入结果，其中弹出值和压入结果都会修改栈指针
          // 容易发现弹出运算数和压入结果这两步让栈大小+1和-1，最终栈大小没有变，而且运算也不依赖于栈，所以可以省去这两次修改
          writeln!(w, "  lw t0, 0(sp)")?; // 从栈顶读出运算数
          let op = match op { Neg => "neg", BNot => "not", LNot => "seqz" };
          writeln!(w, "  {} t0, t0", op)?; // 对运算数进行一元运算，用neg计算负，not计算按位非，seqz计算逻辑非
          writeln!(w, "  sw t0, 0(sp)")?; // 保存回栈顶的位置
        }
        IrStmt::Binary(op) => {
          // 与Unary类似，这里的翻译过程可以做一点简化，不用修改三次栈指针，只用修改一次：最终栈大小-1
          writeln!(w, "  lw t0, 0(sp)")?; // 从栈顶读出右操作数
          writeln!(w, "  lw t1, 4(sp)")?; // 从栈顶"下面一个"读出左操作数，注意栈往下走地址实际上是变高的
          writeln!(w, "  add sp, sp, 4")?; // 栈大小-1
          // 这个match根据op来执行运算，结果保存到t0中，有些运算可以用一条指令完成，有些复杂一点，具体细节可以查阅RISC-V的资料
          match op {
            Add => writeln!(w, "  add t0, t1, t0")?,
            Sub => writeln!(w, "  sub t0, t1, t0")?,
            Mul => writeln!(w, "  mul t0, t1, t0")?,
            Div => writeln!(w, "  div t0, t1, t0")?,
            Mod => writeln!(w, "  rem t0, t1, t0")?,
            Lt => writeln!(w, "  slt t0, t1, t0")?,
            Le => {
              writeln!(w, "  slt t0, t0, t1")?;
              writeln!(w, "  xor t0, t0, 1")?;
            }
            Ge => {
              writeln!(w, "  slt t0, t1, t0")?;
              writeln!(w, "  xor t0, t0, 1")?;
            }
            Gt => writeln!(w, "  slt t0, t0, t1")?,
            Eq => {
              writeln!(w, "  xor t0, t0, t1")?;
              writeln!(w, "  seqz t0, t0")?;
            }
            Ne => {
              writeln!(w, "  xor t0, t0, t1")?;
              writeln!(w, "  snez t0, t0")?;
            }
            // And/Or不能直接用and/or指令完成，因为这两条指令执行的是按位运算，但是语言要求进行逻辑运算
            And => {
              writeln!(w, "  snez t0, t0")?;
              writeln!(w, "  snez t1, t1")?;
              writeln!(w, "  and t0, t0, t1")?;
            }
            Or => {
              writeln!(w, "  or t0, t0, t1")?;
              writeln!(w, "  snez t0, t0")?;
            }
          };
          writeln!(w, "  sw t0, 0(sp)")?; // 保存回栈顶的位置
        }
        IrStmt::LocalAddr(x) => {
          writeln!(w, "  add t0, s0, -{}", (x + 1) * 4)?; // 计算变量x的地址，保存到t0中
          writeln!(w, "  sw t0, -4(sp)")?; // 以下两句将t0的值压入栈中
          writeln!(w, "  add sp, sp, -4")?;
        }
        IrStmt::GlobAddr(x) => {
          writeln!(w, "  la t0, {}", p.globs[*x as usize].0)?; // 计算变量x的地址，保存到t0中
          writeln!(w, "  sw t0, -4(sp)")?; // 以下两句将t0的值压入栈中
          writeln!(w, "  add sp, sp, -4")?;
        }
        IrStmt::Load => {
          // 与Unary类似，这里的翻译过程可以做一点简化，不用修改栈指针
          writeln!(w, "  lw t0, 0(sp)")?; // 从栈顶读出load的地址
          writeln!(w, "  lw t0, 0(t0)")?; // 执行load操作
          writeln!(w, "  sw t0, 0(sp)")?; // 保存回栈顶的位置
        }
        IrStmt::Store => {
          // 与Unary类似，这里的翻译过程可以做一点简化，不用修改三次栈指针，只用修改一次：最终栈大小-1
          writeln!(w, "  lw t0, 0(sp)")?; // 从栈顶读出store的值
          writeln!(w, "  lw t1, 4(sp)")?; // 从栈顶"下面一个"读出store的地址
          writeln!(w, "  sw t1, 0(t0)")?; // 执行store操作
          writeln!(w, "  add sp, sp, 4")?; // 栈大小-1
        }
        IrStmt::Label(x) => writeln!(w, ".L.{}.{}:", f.name, x)?, // label名字的格式是.L.{函数名}.{label编号}
        IrStmt::Bz(x) => {
          writeln!(w, "  lw t0, 0(sp)")?;
          writeln!(w, "  add sp, sp, 4")?; // 以上两句将栈顶的值弹出，存入t0
          writeln!(w, "  beqz t0, .L.{}.{}", f.name, x)?; // 若t0为0，则跳转到对应label，否则继续执行下一句
        }
        IrStmt::Bnz(x) => { // 与IrStmt::Bz完全类似
          writeln!(w, "  lw t0, 0(sp)")?;
          writeln!(w, "  add sp, sp, 4")?;
          writeln!(w, "  bnez t0, .L.{}.{}", f.name, x)?;
        }
        IrStmt::Jump(x) => writeln!(w, "  j .L.{}.{}", f.name, x)?,
        IrStmt::Call(x) => {
          writeln!(w, "  jal {}", p.funcs[*x as usize].name)?;
          writeln!(w, "  sw a0, -4(sp)")?; // a0保存了函数的返回值，但是还不在运算栈中，调用者负责把它压入栈中
          writeln!(w, "  add sp, sp, -4")?;
        }
        IrStmt::Swap => { // 交换栈顶和次栈顶的两个元素，先把它们从各自的位置读出来，再写入对方的位置
          writeln!(w, "  lw t0, 0(sp)")?;
          writeln!(w, "  lw t1, 4(sp)")?;
          writeln!(w, "  sw t0, 4(sp)")?;
          writeln!(w, "  sw t1, 0(sp)")?;
        }
        IrStmt::Pop => writeln!(w, "  add sp, sp, 4")?,
        IrStmt::Ret => {
          writeln!(w, "  lw a0, 0(sp)")?; // 从栈顶读出返回值，不用修改栈指针，因为下面会恢复栈指针
          // 函数的epilogue部分，恢复栈指针，s0，ra。被调用者负责清除栈上的参数
          writeln!(w, "  mv sp, s0")?; // s0的值是函数prologue中设定的，参数都在它下面，令sp = s0就清除了所有参数
          writeln!(w, "  lw s0, -{}(sp)", (f.param_cnt + f.var_cnt + 1) * 4)?;
          writeln!(w, "  lw ra, -{}(sp)", (f.param_cnt + f.var_cnt + 2) * 4)?;
          writeln!(w, "  ret")?; // a0保存了函数的返回值
        }
      }
    }
  }
  Ok(())
}

File: src/codegen.cpp
#include <cassert>
#include <cstdio>
#include <cstdarg>
#include <cstring>
#include "codegen.h"

static bool debug_ = false;
const char* FUNC_EXIT = "function_exit";
const char* ELSE = "else";
const char* THEN = "then";
const char* EXIT = "exit";
const char* BEGIN = "begin";
const char* CONTINUE = "continue";
static FNPtr hot_func;
static NDPtr last_node;
static int label_seq, brk_seq, cont_seq;

void debug(const char* fmt, ...) {
    va_list ap;
    va_start(ap, fmt);
    if(debug_)
        printf(fmt, ap);
}

void push(const char* reg) {
    debug("PUSH %s\n", reg);
    printf("  addi sp, sp, -%d\n", POINTER_WIDTH);
    printf("  sw %s, 0(sp)\n", reg);
}

void pop(const char* reg) {
    debug("POP %s\n", reg);
    printf("  lw %s, 0(sp)\n", reg);
    printf("  addi sp, sp, %d\n", POINTER_WIDTH);
}

inline int var_offset(VarPtr var) {
    return var->is_arg ? var->offset : -(var->offset + 3);
}

// 注意这里不再以 fp 为基准
inline void store(const char* src, const char* addr) {
    printf("  sw %s, 0(%s)\n", src, addr);
}

inline void load(const char* src, const char* addr) {
    printf("  lw %s, 0(%s)\n", src, addr);
}

inline void label(const char* type, int seq) {
    printf(".L.%s.%d:\n", type, seq);
}

inline void jmp(const char* type, int seq) {
    printf("  j .L.%s.%d\n", type, seq);
}

inline void beqz(const char* reg, const char* type, int seq) {
    printf("  beqz %s, .L.%s.%d\n", reg, type, seq);
}

inline void bnez(const char* reg, const char* type, int seq) {
    printf("  bnez %s, .L.%s.%d\n", reg, type, seq);
}

bool lvalue_check(NDPtr node) {
    return node && (node->kind == ND_VAR || node->kind == ND_DEREF);
}

void gen(NDPtr node);

void gen_addr(NDPtr node) {
    VarPtr var;
    switch (node->kind)
    {
    case ND_DECL:
    case ND_VAR:
        var = node->var;
        if(var->is_global) {
            printf("  lui t1, %%hi(%s)\n", var->name);
            printf("  addi t0, t1, %%lo(%s)\n", var->name);
        } else {
            printf("  addi t0, fp, %d\n", var_offset(var) * POINTER_WIDTH);
        }
        push("t0");
        break;
    case ND_DEREF:
        gen(node->lexpr);
        break;
    // 一般而言，type-cast 不影响汇编实际行为
    case ND_TYPE_CAST:
        gen_addr(node->lexpr);
        break;
    default:
        assert(false);
    }
}

void gen_binary(NDPtr node) {
    debug("BINARY\n");
    gen(node->lexpr);
    gen(node->rexpr);
    pop("t1");
    pop("t0");
    switch (node->kind) {
    case ND_ADD:
        printf("  add t0, t0, t1\n");
        break;
    case ND_SUB:
        printf("  sub t0, t0, t1\n");
        break;
    case ND_MUL:
        printf("  mul t0, t0, t1\n");
        break;
    case ND_DIV:
        printf("  div t0, t0, t1\n");
        break;
    case ND_MOD:
        printf("  rem t0, t0, t1\n");
        break;
    case ND_LT:
        printf("  slt t0, t0, t1\n");
        break;        
    case ND_LTE:
        printf("  sgt t0, t0, t1\n");
        printf("  seqz t0, t0\n");
        break;        
    case ND_EQ:
        printf("  sub t0, t0, t1\n");
        printf("  seqz t0, t0\n");
        break;        
    case ND_NEQ: 
        printf("  sub t0, t0, t1\n");
        printf("  snez t0, t0\n");
        break;       
    case ND_LOGAND:
        printf("  snez t0, t0\n");
        printf("  snez t1, t1\n");
        printf("  and t0, t0, t1\n");
        break;
    case ND_LOGOR:
        printf("  or t0, t0, t1\n");
        printf("  snez t0, t0\n");
        break;
    default:
        assert(false);
    }
    push("t0");
    debug("BINARY END\n");
}

void gen(NDPtr node) {
    if(!node) 
        return;
    switch (node->kind) {
    // Statement
    case ND_RETURN:
        debug("RETURN\n");
        gen(node->lexpr);
        pop("a0");
        printf("  j .L.%s.%s\n", FUNC_EXIT, hot_func->name);
        break;
    case ND_DECL:
        debug("DECL\n");
        assert(node->var);
        if(node->var->init) {
            // 赋值的方式完全变了
            gen_addr(node);
            gen(node->var->init);
            pop("t0");
            pop("t1");
            store("t0", "t1");
        }
        break;
    case ND_UNUSED_EXPR:    
        debug("UNUSED\n");
        if(node->lexpr) {
            gen(node->lexpr);
            pop("t0");
        }
        break;
    case ND_IF: {
        debug("ND_IF");
        int seq = label_seq++;
        gen(node->cond);
        pop("t0");
        if (node->els) {
            beqz("t0", ELSE, seq);
            gen(node->then);
            jmp(EXIT, seq);
            label(ELSE, seq);
            gen(node->els);
        } else {
            beqz("t0", EXIT, seq);
            gen(node->then);
        }
        label(EXIT, seq);
        break;
    }
    case ND_BLOCK:
        for (auto n = node->body.begin(); n != node->body.end(); ++n)
            gen(*n);
        break;
    case ND_FOR: {
        int seq = label_seq++;
        // 使用局部变量备份　brk_seq, cont_seq, 便于后续恢复
        int brk = brk_seq;
        int cont = cont_seq;
        // brk_seq／cont_seq　总表示当前 break/continue 语句需要跳转到的序号
        // 在进入一个需要处理 break/continue 的结构时，需要更新。在上一步做了备份。
        brk_seq = cont_seq = seq;
        // 约定 for 语句结构如下
        // for (<init>; <condition>; <inc>)
        //     <statement>
        gen(node->init);
        // begin label，位于判断之前，inc 之后回到这里
        label(BEGIN, seq);
        // 如果没有条件语句，直接跳过这一部分，相当于 true
        if (node->cond) {
            gen(node->cond);
            pop("t0");
            // 判断为否则跳转到出口
            beqz("t0", EXIT, seq);
        }
        gen(node->then);
        // continue label, 位于 inc 语句之前, continue 跳转到这里
        label(CONTINUE, seq);
        if (node->inc)
            gen(node->inc);
        jmp(BEGIN, seq);
        label(EXIT, seq);
        // 恢复　brk_seq, cont_seq，使其仍表示当前 break/continue 语句需要跳转到的序号
        brk_seq = brk;
        cont_seq = cont;
        break;
    }
    case ND_WHILEDO: {
        int seq = label_seq++;
        int brk = brk_seq;
        int cont = cont_seq;
        brk_seq = cont_seq = seq;
        label(CONTINUE, seq);
        gen(node->cond);
        pop("t0");
        beqz("t0", EXIT, seq);
        gen(node->then);
        jmp(CONTINUE, seq);
        label(EXIT, seq);
        brk_seq = brk;
        cont_seq = cont;
        break;
    }
    case ND_DOWHILE: {
        int seq, brk, cont;
        seq = label_seq++;
        brk = brk_seq;
        cont = cont_seq;
        brk_seq = cont_seq = seq;
        label(CONTINUE, seq);
        gen(node->then);
        gen(node->cond);
        pop("t0");
        bnez("t0", CONTINUE, seq);
        brk_seq = brk;
        cont_seq = cont;
        break;
    }
    case ND_BREAK:
        jmp(EXIT, brk_seq);
        break;
    case ND_CONTINUE:
        jmp(CONTINUE, cont_seq);
        break;
    // Expression
    case ND_NUM:
        debug("NUM\n");
        printf("  li t0, %d\n", node->val);
        push("t0");
        break;
    case ND_NOT:
        debug("NOT\n");
        gen(node->lexpr);
        pop("t0");
        printf("  seqz t0, t0\n");
        push("t0");
        break;
    case ND_BITNOT:
        debug("BITNOT\n");
        gen(node->lexpr);
        pop("t0");
        printf("  not t0, t0\n");
        push("t0");
        break;
    case ND_NEG:
        debug("NEG\n");
        gen(node->lexpr);
        pop("t0");
        printf("  neg t0, t0\n");
        push("t0");
        break;
    case ND_VAR:
        assert(node->var);
        gen_addr(node);
        pop("t0");
        load("t1", "t0");
        push("t1");
        break;
    case ND_ASSIGN:
        assert(lvalue_check(node->lexpr));
        gen(node->rexpr);
        gen_addr(node->lexpr);
        pop("t0");
        pop("t1");
        store("t1", "t0");
        push("t1");
        break;
    case ND_TERNARY: {
        debug("TERNARY\n");
        int seq = label_seq++;
        gen(node->cond);
        pop("t0");
        beqz("t0", ELSE, seq);
        gen(node->then);
        jmp(EXIT, seq);
        label(ELSE, seq);
        gen(node->els);
        label(EXIT, seq);
        break;
    }
    case ND_REF:
        gen_addr(node->lexpr);
        break;
    case ND_DEREF:
        gen(node->lexpr);
        pop("t0");
        load("t1", "t0");
        push("t1");
        break;
    case ND_FUNC_CALL: {
        debug("FUNC CALL\n");
        assert(node->func_call);
        int nargs = 0;
        // 计算参数值，注意计算顺序
        // 为了简单将参数全部储存在栈上，所以不需要后续处理
        for (auto a = node->func_call->args.rbegin(); a != node->func_call->args.rend(); ++a) {
            gen(*a);
            ++nargs;
        }
        // 调用目标函数
        printf("  call %s\n", node->func_call->name);
        // 将在栈中的参数弹出
        if(nargs > 0)
            printf("  addi sp, sp, %d\n", nargs * POINTER_WIDTH);
        // 返回值压栈
        push("a0");
        break;
    }
    // 一般而言，type-cast 不影响汇编实际行为
    case ND_TYPE_CAST:
        gen(node->lexpr);
        break;
    default:
        gen_binary(node);
    }
    last_node = node;
}

void gen_text(std::list<FNPtr> &func) {
    printf("  .text\n");
    for (auto f = func.begin(); f != func.end(); ++f) {
        FNPtr fn = hot_func = *f;
        last_node = NULL;
        printf("  .global %s\n", fn->name);
        printf("%s:\n", fn->name);
        // Prelogue
        printf("  addi sp, sp, -%d\n", fn->stack_size);
        printf("  sw ra, %d-4(sp)\n", fn->stack_size);
        printf("  sw fp, %d-8(sp)\n", fn->stack_size);
        printf("  addi fp, sp, %d\n", fn->stack_size);

        for (auto n = fn->stmts->body.begin(); n != fn->stmts->body.end(); ++n)
            gen(*n);
        
        // Missing return
        if(!strncmp(fn->name, "main", 4) && (!last_node || last_node->kind != ND_RETURN)) {
            printf("  li a0, 0\n");
        }
        // Epilogue
        printf(".L.%s.%s:\n", FUNC_EXIT, fn->name);
        printf("  lw fp, %d-8(sp)\n", fn->stack_size);
        printf("  lw ra, %d-4(sp)\n", fn->stack_size);
        printf("  addi sp, sp, %d\n", fn->stack_size);
        printf("  ret\n");
    }
}

void gen_data(std::list<VarPtr> &gvars) {
    printf("  .data\n");
    for(auto v = gvars.begin(); v != gvars.end(); ++v) {
        VarPtr var = *v;
        printf("  .globl %s\n", var->name);
        printf("  .align %d\n", 2);
        printf("%s:\n", var->name);
        if(var->init)
            printf("  .word %d\n", var->init->val);
        else
            printf("  .word %d\n", 0);
    }
}

void codegen(Program* prog, bool debug) {
    debug_ = debug;
    gen_data(prog->gvars);
    gen_text(prog->funcs);
}
